### ISSUE TITLE
GetFieldDescriptorAt result in access violation error

### ISSUE THREAD

**Issue by martijnvanschie** [phaetto/windows-credentials-provider#15]
Hi,

We are working on a custom implementation of a credential provider and have tried your version. this one works for us so we tried to create a clean assembly using this as a template,

When i implement the `GetFieldDescriptotAt` method it results in a access violation error (which seems to be a memory pointer issue). I have copied the exact code from your project but it does not work. Also tried another marshal.

(BTW: I noticed that the default implementation of the interface lacks the `[Out]` directive).

```csharp
        public int GetFieldDescriptorAt(uint dwIndex, [Out] IntPtr ppcpfd)
        {
            Debug.Write("GetFieldDescriptorAt");

            try
            {
                Diagnostics.LogMethodEntry();
                _logger.Debug($"GetFieldDescriptorAt: dwIndex [{dwIndex}].");

                var credential = CredentialsManager.GetCredential(0) as ICredentialDescriptor;
                var descriptor = credential.GetDescriptor((int)dwIndex);
                _logger.Debug($"Returning descriptor of type [{descriptor.cpft}] with label [{descriptor.pszLabel}]");
                int size = Marshal.SizeOf(descriptor);
                IntPtr tempPtr = Marshal.AllocCoTaskMem(size);
                //IntPtr tempPtr = Marshal.AllocHGlobal(size);
                Marshal.StructureToPtr(descriptor, tempPtr, false);
                Marshal.StructureToPtr(tempPtr, ppcpfd, false);
                return HRESULT.S_OK;
            }
            catch (Exception ex)
            {
                _logger.Error(ex, $"{ex.Message}");
                return HRESULT.E_UNEXPECTED;
            }
            catch
            {
                _logger.Error($"Uncaught exception");
                throw;
            }
            finally
            {
                Diagnostics.LogMethodExit();
            }
        }
```

No error occures in this method but the GetBitmap raises an issue as seen in the WinDbg log

```bash
onecoreuap\ds\security\fido\credprov\dll\fidoprovider.cpp(62)\fidocredprov.dll!00007FFD731975AD: (caller: 00007FFD6E7AAC48) ReturnHr(2) tid(1010) 80070057 The parameter is incorrect.
onecoreuap\shell\auth\credprov2fahelper\dll\credprov2fahelper.cpp(80)\CredProv2faHelper.dll!00007FFD7A1C554A: (caller: 00007FFD6E7B3ACF) ReturnHr(1) tid(1010) 800704EC This program is blocked by group policy. For more information, contact your system administrator.
    Msg:[Device unlock policy not configured] 
AdviseGetCredentialCountGetCredentialAtGetFieldDescriptorCountGetFieldDescriptorAtGetDescriptorGetFieldDescriptorAtGetDescriptor(d98.1010): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
credprovhost!_GetBitmapData+0x44:
```

I have no clue why this is occuring. I'm actually testing both your provider and mine from the same solution with the same test client (CredUI) so i'm would at least expect both providers to act the same.

Any thoughts on this.

Regards,
Martijn

**Comment by phaetto** [comment]
Hi there Martijn!

I had to check a C++ implementation to figure out that part before, it is a tricky one.

In the original solution the credentials descriptors are living in the same memory space - it is included in the tile instance.
Is that the same in your implementation?

I can see that you are using the following code:
```csharp
var credential = CredentialsManager.GetCredential(0) as ICredentialDescriptor;
var descriptor = credential.GetDescriptor((int)dwIndex);
```

How does your descriptor array looks? It should be of type `_CREDENTIAL_PROVIDER_FIELD_DESCRIPTOR` to marshal the memory bounds correctly (the actual out type is _CREDENTIAL_PROVIDER_FIELD_DESCRIPTOR**).

Since the error comes from bitmap related methods, how does your `GetBitmapValue` looks like?

**Comment by martijnvanschie** [comment]
Hi,

Thank for looking into this. I created a custom version of a credential manager and the `ICredentialDescriptor` interface is added to the class that also implements `ICredentialProviderCredential2` to extend it. It returns an interface that gives me the `_CREDENTIAL_PROVIDER_FIELD_DESCRIPTOR`.

```csharp
internal interface ICredentialDescriptor
{
    _CREDENTIAL_PROVIDER_FIELD_DESCRIPTOR GetDescriptor(int index);
}
```
I could also use 

````csharp
_CREDENTIAL_PROVIDER_FIELD_DESCRIPTOR descriptor = credential.GetDescriptor((int)dwIndex);
_logger.Debug($"Returning descriptor of type [{descriptor.cpft}] with label [{descriptor.pszLabel}]");
````

The bitmap function is below. Nothing special yet 

````csharp
public virtual int GetBitmapValue(uint dwFieldID, out IntPtr phbmp)
{
    Debug.Write("GetBitmapValue");
    Diagnostics.LogMethodEntry();
    phbmp = IntPtr.Zero;

    try
    {
        return HRESULT.S_OK;
    }
    catch (Exception ex)
    {
        _logger.Error(ex, $"{ex.Message}");
        return HRESULT.E_UNEXPECTED;
    }
    finally
    {
        Diagnostics.LogMethodExit();
    }
}
````
I'm almost thinking of build properties of some kind.
Hoe you have a clue because I don't :)

**Comment by phaetto** [comment]
Actually it seems you have a bug in your GetBitmapValue.

You cannot return a NULL pointer (`IntPtr.Zero`) with return type `HRESULT.S_OK`. That signals the caller that you have a valid reference.

If you want the caller to skip your returned value and treat it as invalid, you have to return `HResultValues.E_NOTIMPL`.

The rest of the code seems correct.

**Comment by martijnvanschie** [comment]
So I tried your suggestion. And although it does sounds plausible it did not work for me.

Also I have a logging statement at the top of the function and this one is not showing in the logging so it seams the call to this method from unmanaged code does not reach the COM object.

Also the access violation message feels like unmanaged code can not, or is not allowed to, access the memory block. I took code from multiple examples and none seam to work.

Strange that all other interop does work. I would expect to not get any interaction with the COM object is there was a general issue with the object.

**Comment by phaetto** [comment]
Can you then maybe post the log that you have as well?

If it would be a discrepancy between transferring pointers, it might mean that you don't compile your project on your correct platform. I assume that you use Windows 10, 64bits, is that true?

**Comment by phaetto** [comment]
I just also noticed that you said you are using the `ICredentialProviderCredential2` interface instead. Have you wired this up correctly on the interfaces?

If you have, you should be following the technical guide from here:
https://github.com/phaetto/windows-credentials-provider/blob/master/Credential_Provider_Technical_Reference.pdf
There are several steps that need to be supported in addition to the v1 interface.

Can you maybe try switching back to ICredentialProviderCredential on your interface and see if that solves your problem?

I know that might not sounds much, but without any form of logging or sample I cannot help much either.

**Comment by martijnvanschie** [comment]
Yes I know it's a tricky one. I basically took a working version and created a new project from scratch. Took out all the overhead and tried to implement the provider step by step. I have quite some experience in programming and this issue is really not making sense so kind of frustrating. Going back to WinDbg has been was long ago :)

I do appreciate the help and I'll see if I can get anything shared but it's a work project so I might need to clean some code before  can do that.

**Comment by martijnvanschie** [comment]
Hi Alexander,

I uploaded some code and testing tool and shared the repo with you. If you have some time could you take a look at it?
Hope you can help me out here :)

Regards,
Martijn

**Comment by phaetto** [comment]
Hey Martijn,

I had some time to give it a look and I can see the following problems with your cred provider:

- COM interfaces are important. Add interface `ISiaasCredentialProvider` and it should support IUnknown. If you haven't been working with COM before, I highly recommend not to start from scratch. 
- Method `GetCredentialAt` returns `HRESULT.S_OK` by setting a null pointer. You really should check all the methods for that pattern - it is dangerous and you can miss it. Better, make some unit tests on those methods.
- SiaasCredentialProviderCredential is not a COM component - did you forget the COM related attributes?

I don't have time to post any changes unfortunately, maybe in the future.

Regards,
Alexander

**Comment by martijnvanschie** [comment]
Hello Alexander.

Really appreciate you looking into this. I do have experience with COM from way back in the days of Classic ASP and COM Interop but a time long forgotten :)

I checked some other implementations and understood that only the class implementing `ICredentialProvider` should be COM visible but I will definitely look into that again.

I will get back to you if I find the solution to this.

Regards,
Martijn

**Comment by martijnvanschie** [comment]
Man, you just saved my life.

It was actually the `GetCredentialAt` method that had the bug. It did indeed always return a null and never set the actual `ICredentialProviderCredential` when found. It works.

Thanks again.

**Comment by phaetto** [comment]
Glad to hear that!

Feel free to create another issue if you have any more problems.
